# Co-Evolution and Epistasis Analysis

Panaroo includes an implementation of the [Spydrpick](https://academic.oup.com/nar/article/47/18/e112/5541093) algorithm for identifying co-evolving genes and genes involved in epistatic interactions. The Spydrpick algorithm makes use of pairwise Mutual Information (MI) values to identify interactions that appear to be signifiantly different from the background distribution.

The Spydrpick algorithm requires either a phylogeny or clustering to control for population structure. A phylogeny can be built form the core genome using [Iqtree](http://www.iqtree.org/). Suitable clusterings can be generated for the pangenome using [Poppunk](https://poppunk.readthedocs.io/en/latest/) or from the core genome using [fastbaps](https://github.com/gtonkinhill/fastbaps).

The spydrpick algorithm can be run using

```
mkdir coevo_results
panaroo-spydrpick -i gene_presence_absence.Rtab -o coevo_results --clusters clusters.csv
```


### Parameters

For a full description of the algorithm and its parameters please see the Spydrpick publication.

```
Identify associations between genes using the Spydrpick algorithm

optional arguments:
  -h, --help            show this help message and exit
  --quantile QUANTILE   the quantile used to determine a threshold for keeping
                        MI values (default=0.9).
  --version             show program's version number and exit

Input/output:
  -i PA_FILE, --input PA_FILE
                        gene presence absence file (.Rtab) generated by
                        Panaroo
  -o OUTPUT_DIR, --out_dir OUTPUT_DIR
                        location of an output directory
  --tree TREE_FILE      phylogeny in newick format for weigting samples to
                        control for population structure
  --clusters CLUSTER_FILE
                        sample clusters for weigting to control for population
                        structure. format: 'sample_name,cluster_id'
```

### Output

The program will produce a file `gene_pa_spydrpick.csv` which includes those gene pairs with MI values above the specified threshold. It includes an 'outlier' column with values


| Outlier Flag |   Description   |
|:------------:|:---------------:|
|       0      |   non-outlier   |
|       1      |     outlier     |
|       2      | extreme outlier |


### Alternative Tools

The output from Panaroo can also be used as input to a number of other co-evolution estimation tools. These include

- [pantagruel](https://github.com/flass/pantagruel)
- [coinfinder](https://github.com/fwhelan/coinfinder)


